include/openthread/tcp.h

```diff
@@ -0,0 +1,589 @@
+ *
+ * @param[in]  aInstance  A pointer to an OpenThread instance.
+ * @param[in]  aSocket    A pointer to an active TCP socket structure.
+ * @param[in]  aArgs      A pointer to a structure of arguments.
+ * @param[in]  aContext   A pointer to application-specific context.
+ *
+ * @retval OT_ERROR_NONE    Successfully opened the socket.
+ * @retval OT_ERROR_FAILED  Failed to open the socket.
+ */
+otError otTcpActiveOpen(otInstance *aInstance, otTcpActiveSocket *aSocket, otTcpActiveOpenArgs *aArgs);
+
+/**
+ * Sets the "connect done" callback on an active TCP socket.
+ *
+ * @param[in]  aInstance             A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket               A pointer to the active TCP socket structure whose callback to set.
+ * @param[in]  aConnectDoneCallback  The new callback.
+ *
+ * @retval OT_ERROR_NONE    Successfully set the callback.
+ * @retval OT_ERROR_FAILED  Failed to set the callback.
+ */
+otError otTcpActiveSetConnectDoneCallback(otInstance *aInstance, otTcpActiveSocket *aSocket, otTcpConnectDone aConnectDoneCallback);
+
+/**
+ * Sets the "send done" callback on an active TCP socket.
+ *
+ * @param[in]  aInstance          A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket            A pointer to the active TCP socket structure whose callback to set.
+ * @param[in]  aSendDoneCallback  The new callback.
+ *
+ * @retval OT_ERROR_NONE    Successfully set the callback.
+ * @retval OT_ERROR_FAILED  Failed to set the callback.
+ */
+otError otTcpActiveSetSendDoneCallback(otInstance *aInstance, otTcpActiveSocket *aSocket, otTcpSendDone aSendDoneCallback);
+
+/**
+ * Sets the "send ready" callback on an active TCP socket.
+ *
+ * @param[in]  aInstance           A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket             A pointer to the active TCP socket structure whose callback to set.
+ * @param[in]  aSendReadyCallback  The new callback.
+ *
+ * @retval OT_ERROR_NONE    Successfully set the callback.
+ * @retval OT_ERROR_FAILED  Failed to set the callback.
+ */
+otError otTcpActiveSetSendReadyCallback(otInstance *aInstance, otTcpActiveSocket *aSocket, otTcpSendReady aSendReadyCallback);
+
+/**
+ * Sets the "receive available" callback on an active TCP socket.
+ *
+ * @param[in]  aInstance                  A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket                    A pointer to the active TCP socket structure whose callback to set.
+ * @param[in]  aReceiveAvailableCallback  The new callback.
+ *
+ * @retval OT_ERROR_NONE    Successfully set the callback.
+ * @retval OT_ERROR_FAILED  Failed to set the callback.
+ */
+otError otTcpActiveSetReceiveAvailableCallback(otInstance *aInstance, otTcpActiveSocket *aSocket, otTcpReceiveAvailable aReceiveAvailableCallback);
+
+/**
+ * Sets the "connection lost" callback on an active TCP socket.
```

Comment:
Should we trigger a connection lost event when the TCP state enters `TIME-WAIT` ?

Reply:
Currently, we don't. I'm open to providing a notification when the TCP state enters `TIME-WAIT`. Would others find this useful?

Reply:
I think the TCP socket should be considered disconnected when enters TIME-WAIT, because it's no longer available for reading or writing. So we should trigger `connection lost` when entering `TIME-WAIT`.

Reply:
That's a good point. But the user also needs another callback when the connection moves from `TIME-WAIT` to `CLOSED`, since the TCP socket can't (or at least shouldn't) be repurposed for a different connection until that happens.

One proposal is to indicate this via the error code. We should probably change it from `otError` to an enumeration that gives the reason (aborted, timed out, closed, time wait). Then a socket that's closed normally would get two callbacks, one when entering `TIME-WAIT` and then another when entering `CLOSED`. What do you think about that solution.

Reply:
I pushed an update implementing this. If we later decide we don't want to do it this way, we can revert the commit and/or update the code again.